### PR TITLE
docs: Update Sphinx to v4.5.0

### LINKED ISSUE #18885: Documentation: Update the version for Sphinx
The version of Sphinx (and other related packages listed in `Documentation/requirements.txt`) is rather outdated. Switching to a newer version could maybe help us benefit from more recent features (I've seen requests for `<details></details>` blocks), or improved Markdown parsing through MyST for the few cases where we use it (generated Markdown documentation).

### PR DESCRIPTION
**Pull request by qmonnet**
This PR updates Sphinx to v4.5.0. A number of dependencies are updated as well. In particular, the Alpine image for the builder for the docs is updated.

This comes with a number of minor adjustments, please refer to individual commit descriptions for details.

Note that the last commit update the theme for the documentation: this is required after the Sphinx upgrade. The current version of this commit points to [my own clone of the theme](https://github.com/qmonnet/sphinx_rtd_theme/commits/wip/rebase) (with our changes rebased on latest upstream), because this is still under test. If everything goes well this new version should be pushed to a dedicated branch on Cilium's fork of the theme, and the last commit of this PR should be updated to point to it. The last commit, updating the CI image for building the documentation, is also to be removed after the main image has been updated.

Motivations:
- More add-ons available, e.g. `<details>` blocks (and Mermaid rendering?)
- No more `npm install` (broken on M1 Macs) for using the theme
- More configuration options in Sphinx and add-ons, some of which we wanted to use in the past
- Slightly better context information provided on build errors
- More up-to-date software and Docker images

Fixes: #18885

### CODE REVIEW

**Review by michi-covalent — APPROVED** [review]
🚢

**Review by twpayne — APPROVED** [review]

**Comment by qmonnet** [comment]
Updating the `docs-builder` image used by the CI workflow would be smoother if we didn't rely on `:latest`, but used a pinned version instead, so I opened https://github.com/cilium/cilium/pull/19356.

**Review by sayboras — APPROVED** [review]

**Comment by qmonnet** [comment]
Note: Ideally, I also need reviews [on the rebasing of the theme we use for the docs](https://github.com/qmonnet/sphinx_rtd_theme/commits/wip/rebase). I can create a PR for it but I'm not sure against which branch. Which also brings up the discussion: Where should this update of the theme go? Should I:
1. force-push on the `0.7` branch of the related repository, so we keep using the same branch? Probably not, because we won't have a chance to revert if something breaks.
2. update and push to the `master` branch of that repository? The current `master` is outdated and unused. But similarly, if we rebase again in the future, we'll need backup branches somehow?
3. create a new version branch (`0.8` or even `1.x` to show that it's on top of `sphinx_rtd_theme` `1.0.0`), and switch to that one?

Option 3. sounds best to me. We could also update the `master` branch on https://github.com/cilium/sphinx_rtd_theme so that it just follows upstream, and doesn't contain outdated commits of ours; or we could delete it entirely?

File: .github/workflows/documentation.yaml
Comment (joestringer) [inline .github/workflows/documentation.yaml]: I assume you'll revert this before merging? Do you need assistance in pushing a new version?

File: Documentation/check-build.sh
Comment (joestringer) [inline Documentation/check-build.sh:36]: The commit message doesn't seem to explicitly call this out, but does this mean that these previous extraneous warnings are no longer present or no longer trigger a docs build failure? Awesome!

File: Documentation/requirements.txt
Comment (joestringer) [inline Documentation/requirements.txt]: One more reference to bump before we merge.

**Review by joestringer — CHANGES_REQUESTED** [review]
I filtered just to my own codeowners and reviewed, but it all looks good :+1: 

Let's coordinate to make sure that we continue to rely on shared community docker images / repos before we merge this.

File: .github/workflows/documentation.yaml
Comment (qmonnet) [reply .github/workflows/documentation.yaml]: Yes and yes. I don't have permission to push so I'll need your help once reviews are good and we consider the PR in good shape. Once pushed, I can update this commit to point to the specific image (since we merged https://github.com/cilium/cilium/pull/19356).

File: Documentation/check-build.sh
Comment (qmonnet) [reply Documentation/check-build.sh:36]: Commit log has

```
  [...] Since we touch the script,
  remove filtering expressions which do not seem to catch any existing
  warning.
```

Yes, none of which are present anymore as far as I can tell, so we can just remove this part :)

File: Documentation/requirements.txt
Comment (qmonnet) [reply Documentation/requirements.txt]: Absolutely. This was conditioned by the discussion above - Since we agreed on creating a `1.x` branch on that repo, I'll go ahead and do that (if I have the permissions, haven't checked yet). Then I'll come back to this PR and update this line.

File: Documentation/check-build.sh
Comment (joestringer) [reply Documentation/check-build.sh:36]: Ah right, I glanced through and missed this. yay :tada:

File: Documentation/requirements.txt
Comment (qmonnet) [reply Documentation/requirements.txt]: Now updated to point to a new branch https://github.com/cilium/sphinx_rtd_theme/tree/v1.0. This branch is a rebase of branch `v0.7`, on top of the latest upstream commits. Some clean-up (commit squashing, drop of indentation-only changes, etc.) were done during that rebase, to facilitate potential rebases in the future. Our four latest commits were omitted (because they came after I did that rebase, and the last two are reverts of the former two, so they're all a no-op).

The `master` branch on that repo is now identical to the one in the upstream repository (https://github.com/readthedocs/sphinx_rtd_theme).

File: .github/workflows/documentation.yaml
Comment (qmonnet) [reply .github/workflows/documentation.yaml]: Now updated to point to the image you pushed, thanks!

**Review by joestringer — APPROVED** [review]

**Comment by ldelossa** [comment]
@qmonnet is it out of scope for this PR to add `mermaid` support as well? 

This would make it possible for my to migrate some of the BGP docs immediately. 

If not, totally fine we can do it in a subsequent PR.

**Review by ldelossa — APPROVED** [review]

**Pull request merged**
